Bluetooth: Mesh: Add OOB Public Key support for provisionee role

PavelVPV opened the pull request:
This commit allows an unprovisioned device to exchange its public key using
out-of-band techology. See MeshPRFv1.0.1, table 5.19 and section 5.4.2.3.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
Mesh uses controller to generate DH key. I guess, to support static OOB public key, DH key generation should be moved up to mesh.

PavelVPV commented:
Clarification on how it works (perhaps I should make this clear in API). The key pair can be either dynamic or static.

To use dynamic key:
1. Set `pub_key_set` to `BT_MESH_PUB_KEY_OOB`;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L115
2. Call `bt_mesh_prov_local_pub_key_cb_register` to retrieve generated public key before provisioning starts to send it OOB;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L304

Dynamic pair re-generated on every provisioning attempt (as without OOB Public Key). The user callback will be called every time when a new public key is generated. Perhaps, this won't happen, if the key pair is re-generated implicitly by BLE stack before the provisioning happens.

To use static key pair:
1. Set `pub_key_set` to `BT_MESH_PUB_KEY_OOB`;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L115
2. Initialise `public_key` and `private_key` pointers before provisioning is started;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L117-L120

In case of static key pair, DH key generated directly in provisionee module instead of using controller: https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L149-L150
https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L346-L348
https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L312-L338

trond-snekvik commented on include/bluetooth/mesh/main.h:
```diff
@@ -105,6 +111,14 @@ struct bt_mesh_prov {
+	/** Pointer to Public Key for static OOB Public Key */
+	const uint8_t *public_key;
+	/** Pointer to Private Key for static OOB Public Key */
+	const uint8_t *private_key;
```
This is different from the provisioner side of the OOB mechanism, which uses a function. It would be nice to use the same pattern (if not the same mechanism), but I don't see any obvious solution to this. Any suggestions?

PavelVPV replied:
I guess in case of provisioner the idea is that public key changes from provisionee to provisionee, while for provisionee it's kind of static, which doesn't make sense to use a function.

trond-snekvik replied:
That's true, yeah. I might be looking for a pattern where there is none :slightly_smiling_face:

trond-snekvik reviewed (changes requested):
Is there a usecase that we need to support with the dynamic OOB key, or could we just stick with the static OOB key for those, and require users to set them beforehand? AFAICT, it's possible to achieve the same things with this approach as well, although it moves some extra work into the app.

This could remove the need for the `pub_key_type` field and the pubkey callback, and would make the mixed dh_key_gen mechanism more clear cut and easier to follow. The key generation part was already the most intricate of the provisioning state machine, so it'd be nice if we can avoid adding another dimension to it.

I agree that this needs some conceptual documentation, even if we remove the dynamic OOB key mechanism. E.g.:
- When to use OOB public key
- Requirements for the key material
- A short snippet showing how, since it's not used in any samples

PavelVPV commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -311,7 +368,24 @@ static void prov_pub_key(const uint8_t *data)
+		if (atomic_test_bit(bt_mesh_prov_link.flags, OOB_PUB_KEY_STATIC)) {
+			local_pk = bt_mesh_prov->public_key;
+		} else {
+			/* If we got to this point, Public Key is already generated. */
```
When OOB Public key type is chosen by provisioner, it (by spec, figure 5.15) sends its public key after it reads out provisionee's public key over OOB. Although, provisionee should probably fail here if public key is not ready.

LingaoM commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -724,8 +717,8 @@ int bt_mesh_prov_remote_pub_key_set(const uint8_t public_key[64])
-	memcpy(&bt_mesh_prov_link.conf_inputs[81], public_key, 32);
-	memcpy(&bt_mesh_prov_link.conf_inputs[81 + 32], &public_key[32], 32);
+	sys_memcpy_swap(&bt_mesh_prov_link.conf_inputs[81], public_key, 32);
+	sys_memcpy_swap(&bt_mesh_prov_link.conf_inputs[81 + 32], &public_key[32], 32);
```
Can you put this in a separate commit?

LingaoM replied:
https://github.com/zephyrproject-rtos/zephyr/blob/efccfcbfc91501cac2ede0541a48c338f2b0cae4/subsys/bluetooth/mesh/provisioner.c#L414-L415

Have done the same thing here?

LingaoM commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -280,9 +306,35 @@ static void prov_dh_key_cb(const uint8_t dhkey[32])
+
+	sys_memcpy_swap(private_key_be, bt_mesh_prov->private_key, 32);
+
+	ret = uECC_valid_public_key(remote_pk_be, &curve_secp256r1);
+	if (ret < 0) {
+		BT_ERR("public key is not valid (ret %d)", ret);
+		ret = TC_CRYPTO_FAIL;
+	} else {
+		ret = uECC_shared_secret(remote_pk_be, private_key_be,
+					 bt_mesh_prov_link.dhkey, &curve_secp256r1);
+	}
```
If we put ECC in Mesh, do we need to re-evaluate Stack resource occupation?
https://github.com/zephyrproject-rtos/zephyr/blob/efccfcbfc91501cac2ede0541a48c338f2b0cae4/subsys/bluetooth/mesh/Kconfig#L261

LingaoM commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -291,16 +343,21 @@ static void prov_dh_key_gen(void)
+		sys_memcpy_swap(remote_pk_le, remote_pk, 32);
+		sys_memcpy_swap(&remote_pk_le[32], &remote_pk[32], 32);
+
+		if (bt_dh_key_gen(remote_pk_le, prov_dh_key_cb)) {
```
Is it necessary to use the existing structure and add specific functions so that the bt controller can calculate the given private/public key ?

LingaoM commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -280,9 +306,35 @@ static void prov_dh_key_cb(const uint8_t dhkey[32])
+	}
 
-	send_pub_key();
+	if (ret == TC_CRYPTO_FAIL) {
```
```suggestion
	if (ret != TC_CRYPTO_SUCCESS) {
```

LingaoM commented on include/bluetooth/mesh/main.h:
```diff
@@ -272,6 +286,23 @@ int bt_mesh_input_string(const char *str);
+ *
+ * A new key is generated after each provisioning attempt.
+ *
+ * Used only by Provisionee.
```
According to https://github.com/zephyrproject-rtos/zephyr/pull/28457#discussion_r501068834 .
Should it be:
```suggestion
 * Used only by device provisioning.
```

LingaoM commented on include/bluetooth/mesh/main.h:
```diff
@@ -272,6 +286,23 @@ int bt_mesh_input_string(const char *str);
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+int bt_mesh_prov_local_pub_key_cb_register(void (*cb)(const uint8_t *key));
```
Why implement such a function?

This function only provides public key capabilities, and does not have private key information,
so even if the App has a public key, it's still cannot use it for `prov->private_key`. 

https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L320

So public/private key can only be static.

LingaoM replied:
My suggestion is that the App layer defines its own way of generating public and private keys. The protocol stack does not provide capabilities. Therefore, this registration can be omitted.

LingaoM commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -13,6 +13,11 @@
+#include <tinycrypt/constants.h>
+#include <tinycrypt/utils.h>
+#include <tinycrypt/ecc.h>
+#include <tinycrypt/ecc_dh.h>
```
https://github.com/zephyrproject-rtos/zephyr/blob/38eaabc9c75c9a0932bbd272952dc81a2f2939ef/subsys/bluetooth/mesh/Kconfig#L261-L266

If we want to implement ECC in Mesh, do we need to re-evaluate the resource occupation of ADV Stack?

trond-snekvik replied:
I don't think so, this does not add calls to the public key generation from the send callbacks

LingaoM replied:
`prov_pub_key` --> `prov_dh_key_gen` --> `static_dh_key_gen`

trond-snekvik replied:
`prov_pub_key` comes from the scanner thread though

LingaoM replied:
In fact, should it be `BT_RX_STACK_SIZE` ?

PavelVPV replied:
Thread usage is the same in both cases: with and without OOB public key:
```
Thread analyze:         
 BT Mesh adv         : STACK: unused 248 usage 520 / 768 (67 %); CPU: 0 %
 BT RX               : STACK: unused 864 usage 1184 / 2048 (57 %); CPU: 2 %
 BT RX pri           : STACK: unused 296 usage 152 / 448 (33 %); CPU: 0 %
 BT ECC              : STACK: unused 256 usage 888 / 1144 (77 %); CPU: 1 %
 BT TX               : STACK: unused 272 usage 368 / 640 (57 %); CPU: 0 %
 sysworkq            : STACK: unused 7048 usage 1144 / 8192 (13 %); CPU: 0 %
 shell_uart          : STACK: unused 1616 usage 432 / 2048 (21 %); CPU: 0 %
 idle 00             : STACK: unused 264 usage 56 / 320 (17 %); CPU: 95 %
```

alxelax commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -280,9 +306,35 @@ static void prov_dh_key_cb(const uint8_t dhkey[32])
+	dh_key_gen_complete();
+}
+
+static void static_dh_key_gen(void)
```
Shouldn't it be better to keep all ECC dependent algorithms in crypto.c\.h?

PavelVPV replied:
It could be moved, but would it make a difference?

alxelax replied:
at the current moment no. However, if someone :) decides to use another security toolbox or to move this code to the crypto memory area it would be done easier.

PavelVPV committed:
Bluetooth: Mesh: Add OOB Public Key support for provisionee role

This commit allows an unprovisioned device to exchange its public key
using out-of-band techology (see MeshPRFv1.0.1, table 5.19 and section
5.4.2.3).

For in-band public key exchange, the mesh stack uses HCI commands to
generate public and private keys, and DH key. This, however, doesn't
work for OOB public key exchange since there is no command to generate
DH key with a private key provided by an application. Therefore, this
commit adds direct usage of TinyCrypto into the mesh stack for DH key
generation for OOB public key support.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
Bluetooth: Mesh: Fix setting remote public key in provisioner

This aligns provisioner and provisionee APIs in terms of endianess
of public key provided by an application.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
Thanks everyone for the comments! I hope they are resolved now.

LingaoM reviewed (approved):
Great work. LGTM!

jhedberg commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -92,8 +97,8 @@ static void prov_invite(const uint8_t *data)
 
-	/* Public Key Type, Only "No OOB" Public Key is supported */
-	net_buf_simple_add_u8(&buf, PUB_KEY_NO_OOB);
+	/* Public Key Type */
+	net_buf_simple_add_u8(&buf, !!bt_mesh_prov->public_key_be);
```
I know it'll be more lines of code, but I think it would be better to explicitly map to one of the allowed Public Key type values to make it clear that you're translating from some internal implementation detail (NULL pointer or not) to a valid provisioning protocol value.

So something like:
```
if (bt_mesh_prov->public_key_be == NULL) {
        net_buf_simple_add_u8(buf, PUB_KEY_NO_OOB);
} else {
        net_buf_simple_add_u8(buf, PUB_KEY_OOB);
}

trond-snekvik commented on include/bluetooth/mesh/main.h:
```diff
@@ -105,6 +105,23 @@ struct bt_mesh_prov {
+	/** Pointer to Private Key in big-endian for OOB public key type support.
+	 *
+	 * Remember to enable :c:option:`CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY`
+	 * when initializing this parameter.
```
I'd suggest adding a warning in `bt_mesh_prov_enable` if a keypair is set, but OOB pubkey is disabled. E.g.
```c
	if (!IS_ENABLED(CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY) && bt_mesh_prov->public_key_be) {
		BT_WARN("OOB Public key disabled");
	}
```

PavelVPV replied:
I was thinking about that, but I decided that one warning in `prov_device.c` (https://github.com/zephyrproject-rtos/zephyr/blob/04e81153e066577d39a96a209d9444182cd03aa3/subsys/bluetooth/mesh/prov_device.c#L145) is enough. This is something that you do once since it's compile time option.

trond-snekvik replied:
Alright, sounds fair enough :+1:

trond-snekvik commented on include/bluetooth/mesh/main.h:
```diff
@@ -105,6 +105,23 @@ struct bt_mesh_prov {
+	 * Remember to enable :c:option:`CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY`
+	 * when initializing this parameter.
+	 *
+	 * Must be used together with :c:member:`struct bt_mesh_prov.public_key_be.`
```
Perhaps we could return an error from `bt_mesh_prov_enable` if only one is non-null? E.g.
```c
	if (!!bt_mesh_prov->public_key_be != !!bt_mesh_prov->private_key_be) {
		BT_WARN("Both public and private key must be set");
		return -EINVAL;
	}
```

PavelVPV replied:
There is the warning here: https://github.com/zephyrproject-rtos/zephyr/blob/04e81153e066577d39a96a209d9444182cd03aa3/subsys/bluetooth/mesh/prov_device.c#L145

trond-snekvik commented on include/bluetooth/mesh/main.h:
```diff
@@ -105,6 +105,23 @@ struct bt_mesh_prov {
+	 * Remember to enable :c:option:`CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY`
+	 * when initializing this parameter.
+	 *
+	 * Must be used together with :c:member:`struct bt_mesh_prov.private_key_be.
```
I'm not 100% sure, but I believe this doesn't work with `struct blahblah`, it needs to just be `blahblah`
